rfcs/0010-custom-system-uris.md

```diff
@@ -0,0 +1,325 @@
+  don't waste time writing responses that don't deliver real value.
+For any such not-needed section, simply write in "N/A".
+-->
+
+The question being considered by this RFC is this:
+  what base URI/URL should the BFD API use in its data payloads for custom/non-standard FHIR `system`s?
+This RFC makes the recommendation,
+  as explained and justified in later sections,
+  that BFD should stick with a base of
+  `https://bluebutton.cms.gov/resources/` for all custom URIs.
+
+Before we go further, a quick note on terminology:
+  this RFC talks a lot about
+  [URI](https://datatracker.ietf.org/doc/html/rfc3986)s and
+  [URL](https://datatracker.ietf.org/doc/html/rfc1738)s,
+  which are _similar_ but nevertheless distinct concepts.
+Briefly, a URI is a "magic string" in a specific format that uniquely identifies something.
+Whereas a URL is also a "magic string" but one that web browsers and other software applications can use to retrieve something.
+Often, URLs start with "http://", indicating that a web browser can likely open it.
+In the vein of "a square is a rectangle but not all rectangles are squares",
+  a URL is a URI but not all URIs are URLs.
+
+BFD's data payloads include a number of custom URIs for FHIR
+  [Coding.system](https://www.hl7.org/fhir/datatypes-definitions.html#Coding.system)s,
+  [Identifier.system](https://www.hl7.org/fhir/datatypes-definitions.html#Identifier.system)s,
+  and other such data types.
+These URIs are serve two functions:
+Firstly, they are "magic strings" that end user applications will use to select the pieces of our data payload that their applications care about.
+Second, they can also be treated as URLs that provide developer documentation, e.g. data field definitions.
+While many of the elements in our data payload can use industry-standard URIs for industry-standardized data,
+  a large percentage of our data elements/codings are custom to CMS and thus must use custom URIs.
+
+As an example, consider the following sample `Patient` resource payload from BFD:
+  <https://github.com/CMSgov/beneficiary-fhir-data/blob/master/apps/bfd-server/bfd-server-war/src/test/resources/endpoint-responses/v2/patientRead.json>.
+If an application wanted to select the `Patient.identifier` entry that had a beneficiary's MBI,
+  they would search for the entry with a `system` value of `http://hl7.org/fhir/sid/us-mbi`.
+HL7, a standards body, has said that `system` should be used for all MBIs,
+  in all FHIR-compliant APIs, and so BFD does not need a custom URI there.
+If, however, an application wants to select the `Patient.identifier` entry that contains a beneficiary's CCW beneficiary ID,
+  they would search for the entry with a `system` value of `https://bluebutton.cms.gov/resources/variables/bene_id`.
+The "CCW beneficiary ID" field is only used within/by CMS,
+  so no standards body has assigned it a standard URI,
+  and BFD must provide a custom URI for it.
```

Comment:
This is a great way to explain it.  I think there is also a third function that these uri's/url's provide too, which is to allow for validation of the coding.code contained within the FHIR resource against the coding.system url.  There is typically a .json addition to the URL's webpage that a FHIR validator can use to perform these validations for coding systems/codes.

Reply:
Isn't offering a terminology service (at some point in the future) an alternative? Or would that need to be hosted at the same URLs?

Reply:
Typically it is at the same URL, as that's how the validator knows 'where' to get the value set to compare the code against, if that makes sense.  There may be a way to 'redirect' to a separate url that has that, but I'm not sure how at this point.  Providing this functionality 'could' be a future endeavor though, as I think they only show up as warnings or information within the validators response.

Reply:
I don't think we're doing anything here that _precludes_ supporting that validation in the future, so I'm going to mark this as resolved for now. If I'm misunderstanding things, though, **please** comment and unresolve this.